### PR TITLE
Support binary payloads and responses, and non-UTF8 string payloads for Python functions

### LINKED ISSUE #205: Support binary payloads and responses, and non-UTF8 string payloads
The Python templates currently don't support binary payloads and responses, whereas other templates do (e.g. image-handling functions in store). Much like #203 and #159 for Java, it would be nice to add this capability to the Python templates. Additionally, string payloads are assumed to be UTF-8 encoded, even if content-type signals differently.

## Expected Behaviour

When passing binary payload, don't decode to string. When passing string payload, respect content-type encoding.
When returning binary file, don't encode as string.

## Current Behaviour

UnicodeDecodeError when passing binary payload, incorrect decoding when passing non-UTF-8 string payload. Binary responses get stringified.

## Possible Solution

I've got this fixed and working for more than a year in my custom template, happy to send a PR

## Context

Using audio files as payload

## Your Environment

* Docker version `docker version` (e.g. Docker 17.0.05 ):
All
* Are you using Docker Swarm or Kubernetes (FaaS-netes)?
Swarm
* Operating System and version (e.g. Linux, Windows, MacOS):
Linux, MacOS
* Link to your project or a code example to reproduce issue:

### PR DESCRIPTION
**Pull request by jpauwels**
closes #205 and replaces #207 

## Description
In addition to handling binary input as in #207, this PR also handles binary output, properly encodes returned Python dicts as compliant JSON (double quotes instead of single quotes) and can handle all text encodings. All this processing is controlled by the headers according to web standards.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## Which issue(s) this PR fixes 
Fixes #205 

## How Has This Been Tested?
Deployed on an internal project that takes audio files/JSON as payload and response, running on Ubuntu and macOS.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Version change (see: Impact to existing users)

## Impact to existing users
UTF8 text payloads (the only ones supported so far) are passed as before.

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by LucasRoesler** [comment]
I think this looks good, @alexellis what do you think?

**Comment by alexellis** [comment]
I haven't had a chance to look at it yet.

It may be better just to use the existing support for binary responses in the HTTP/flask templates that I added whilst building out the LF training course on serverless

https://github.com/openfaas-incubator/python-flask-template/commit/9fb55274a5fed0840122a9eb88466d4742fd901d

File: template/python3/index.py [hunk 1]
```diff
@@ -3,19 +3,37 @@
...
+    # decode JSON and JSON-LD to dictionary
+    elif content_type.endswith('json'):
+        st = json.load(st)
+    # otherwise keep bytes
+
     ret = handler.handle(st)
-    if ret != None:
-        print(ret)
+
+    # encode response to JSON if necessary
```
Comment (alexellis) [inline template/python3/index.py:32]: This part seems fine in principle, but be aware that you also need to set the mime type in the function's deployment. It cannot be dynamic with the classic watchdog.

File: template/python3/index.py [hunk 2]
```diff
@@ -3,19 +3,37 @@
...
-        buf += line
-        if line == "":
-            break
-    return buf
-
 if __name__ == "__main__":
-    st = get_stdin()
+    st = sys.stdin.buffer.read()
+    # decode text to string
+    content_type, type_options = parse_header(os.getenv('Http_Content_Type', ''))
```
Comment (alexellis) [inline template/python3/index.py:14]: You don't mention `application/x-www-form-urlencoded` in the PR description?

**Comment by alexellis** [comment]
Quite a few changes and output from testing. It would be good to see a sample GitHub repo where you have deployed samples from each of the scenarios that are supported in the response writing code.

**Comment by alexellis** [comment]
@jpauwels we didn't hear back from you after providing feedback, so I'm going to assume you're no longer interested in seeing this through.

For anyone else landing here, do check out the python3-flask/http templates in the store. These are in a separate repository in this organisation.

**Pull request closed without merging**